typescript/schedules.md

```diff
@@ -0,0 +1,161 @@
+- gRPC API:
+  - [methods](https://github.com/temporalio/api/blob/799926c86eb13d8a9717d3561ab9b0df43796c06/temporal/api/workflowservice/v1/service.proto#L328-L370)
+  - [`request_response.proto`](https://github.com/temporalio/api/blob/799926c86eb13d8a9717d3561ab9b0df43796c06/temporal/api/workflowservice/v1/request_response.proto#L821-L957)
+  - [`schedule/v1/message.proto`](https://github.com/temporalio/api/blob/master/temporal/api/schedule/v1/message.proto)
+  - [`ScheduleOverlapPolicy` enum](https://github.com/temporalio/api/blob/master/temporal/api/enums/v1/schedule.proto)
+
+## TS API
+
+```ts
+import { ScheduleClient, ScheduleOverlapPolicy } from '@temporalio/client'
+import { addWeeks } from 'date-fns'
+import { myWorkflow } from './workflows'
+
+const client = new ScheduleClient()
+
+const schedule = await client.create({
+  id: 'schedule-biz-id',
+  spec: {
+    // every hour at minute 5
+    intervals: [{
+      every: '1h',
+      at: '5m',
+    }],
+    // every 20 days since epoch at day 2
+    // intervals: [{
+    //   every: '20d',
+    //   at: '2d'
+    // }]
+    exclude: [{
+      // skip 11:05 pm
+      hour: 23,
+      minute: 5,
+    }],
+    endAt: addWeeks(new Date(), 4), 
+    jitter: '30s',
+    timezone: 'US/Eastern',
+  },
+  action: {
+    startWorkflow: {
+      workflowId: 'wf-biz-id',
+      type: myWorkflow,
+      args: ['sorry this is the only thing reused, chad 😄'],
+      // ...WorkflowOptions + header - followRuns
+      // https://typescript.temporal.io/api/interfaces/client.WorkflowOptions
+    },
+  },
+  policies: {
+    overlap: ScheduleOverlapPolicy.BUFFER_ONE,
+    catchupWindow: '2m',
+    pauseOnFailure: true,
+  },
+  state: {
+    note: 'started schedule',
+    paused: true,
+    limitedActions: 10,
+  },
+  patch: {
+    triggerImmediately: true,
+    backfill: [
+      {
+        start: new Date(),
+        end: new Date(),
+        overlap: ScheduleOverlapPolicy.ALLOW_ALL,
+      },
+    ],
+    pause: true, // redundant with state.paused above (can use either)
+  },
+  memo,
+  searchAttributes,
+})
+
+const schedule = client.getHandle('schedule-biz-id')
+
+const scheduleDescription = await schedule.describe()
+
+const matchingStartTimes = await schedule.listMatchingTimes({ start: new Date(), end: new Date() })
+
+await schedule.update({
+  spec,
+  action,
+  policies,
+  state,
+  conflictToken: scheduleDescription.conflictToken,
+})
+
+await schedule.patch({
+  triggerImmediately: true,
+  backfill,
+  unpause: true,
+})
+
+await schedule.delete()
+
+const { schedules, nextPageToken } = await client.list({
+  pageSize: 50,
+  nextPageToken: 'base64',
+})
+```
+
+### Higher-level client
+
+```ts
+import { Client } from '@temporalio/client'
+
+const client = new Client(options)
+
+client.schedule.create()
+client.workflow.start() 
+client.asyncCompletion.heartbeat()
+client.operator.addSearchAttributes()
+
+interface ClientOptions {
+  dataConverter?: DataConverter;
+  interceptors?: {
```

Comment:
Is there some form of abstract "interceptor base" class like in other languages that by default just sends each thing to the next interceptor? Or do we require people manually implement every method in an interceptor? If you made proper abstract class interceptors, one wonders if this should just be `ClientInterceptor[]`, and `ClientInterceptor` interface just extends the others and there's a base that implements it with default methods.

Reply:
All interceptor methods in TS are optional. No need for an ABC.
I wouldn't merge the interceptors, the WorkflowClientInterceptor method names are already too short and it's too late to change now.
I don't think there's a good reason to add ScheduleClientInterceptors unless we want to inject workflow headers with it.